request_idempotency: Build basics of idempotency logic.

abdelrahman725 opened the pull request:
Fixes #31219

Implement a generic approach used to ensure idempotency for non-idempotent request (e.g. POST).

Currently only used for messsage sending.

timabbott commented on zerver/lib/exceptions.py:
```diff
@@ -801,3 +801,16 @@ class SlackImportInvalidFileError(Exception):
+    @staticmethod
+    @override
+    def msg_format() -> str:
+        return _("Conflict error: {reason}")
```
Usually you don't want to do this sort of construction for internationalization; `reason` still also needs to be tagged for translation. Better would typically be to just have each `LockedError` supply a complete translated string, and have this class just provide the status code.

abdelrahman725 replied:
Sounds correct

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -857,6 +863,101 @@ def get_active_presence_idle_user_ids(
+        "has_link",
+        "is_channel_message",
+        "search_tsvector",
+    ]
```
It'd be good to figure out how to not have to hardcode this list. If you forgot a field here, does the test fail? 

I bet Django internals offer a way to ask for the SQL you need.

abdelrahman725 replied:
Yes, it's possible to generate those fields dynamically

timabbott replied:
I think there's bigger issue here. How can we verify that creating a model object with this method produces the same `Message` row in Postgres as doing so with the `bulk_create` code path?

Django's real `bulk_create` has a bunch of logic in functions like  `_prepare_related_fields_for_save`.

abdelrahman725 replied:
> Yeah. For instance, I think this will not apply `default` values, which are applied in Django, not the database?

No. The default values are already applied to the `Message` instnace, and I get those values in:

```python
column_to_value: dict[str, Any] = {
    col_name: getattr(message, col_name) for col_name in field_names
}
```

abdelrahman725 commented on zerver/lib/custom_zulip_transaction.py:
```diff
@@ -0,0 +1,106 @@
+    This decorator expects the decorated function
+    to optionally receive the idempotency_key as one of its arguments.
+
+    idempotent_work: what type of work are we wrapping, e.g. sending message.
+    This is helpful to know the type of data do_work returns.
```
This should be removed

showell commented on web/src/channel.ts:
```diff
@@ -22,6 +22,17 @@ type AjaxRequestHandlerOptions = Omit<JQuery.AjaxSettings, "success"> & {
+// A generic idempotency manager for any future request we want to make it idempotent.
+// Passed as HTTP header (if defined) to $.ajax().
+// success and error are not used yet but are kept in case
+// we decide to handle each case separately.
```
If success and error are not used yet, then I would just omit them from this PR.  You can still comment future directions, but I wouldn't introduce dead code here.

showell commented on zerver/actions/message_send.py:
```diff
@@ -858,11 +863,118 @@ def get_active_presence_idle_user_ids(
+            msg_cols=SQL(", ").join(map(Identifier, field_names)),
+            values_placeholders=SQL(", ").join(map(Placeholder, field_names)),
+            message_pk=Identifier(Message._meta.pk.name),
+        )
```
Insofar as you think this code will be reusable outside the context of handling Message rows in particular, I think you can optimistically make some things generic.  For example, I would replace message_table and msg_cols and message_pk with just `table` and `cols` and `message_pk` as the format keys.  And I would introduce local vars:

~~~ py
table = Identifier(Message._meta.db_table)
pk = Identifier(Message._meta.pk.name)

// ...
    ..format(
        table=table,
        pk=pk,
        # ...
    )

showell commented:
LGTM. I didn't do a thorough review, but my previous comments have been addressed to my satisfaction.

timabbott commented on zerver/models/messages.py:
```diff
@@ -702,6 +702,29 @@ def __str__(self) -> str:
+    any http request having the possibility of being wrongly replayed.
+    """
+
+    idempotency_key = models.UUIDField(unique=True, db_index=True)
```
Let's add `realm` and `user` columns to this table, so it can be sharded with the realm, and make the unique index be on `(realm, user, idempotency_key)`. 

That would be more in line with our security model, which isolates realms from each other. And this functionality should isolate users from each other too.

timabbott commented on zerver/lib/custom_zulip_transaction.py:
```diff
@@ -0,0 +1,187 @@
+            )
+
+
+def zulip_transaction(
```
I think you want a name like `idempotent_request_transaction` that clarifies what's special about this transaction in the name.

timabbott commented on zerver/lib/custom_zulip_transaction.py:
```diff
@@ -0,0 +1,187 @@
+from collections.abc import Callable
```
I think calling this `zerver/lib/idempotent_request.py` is probably a much better name.

timabbott commented on zerver/tests/test_message_send.py:
```diff
@@ -1508,6 +1508,66 @@ def test_guest_user(self) -> None:
         result = self.api_post(sender, "/api/v1/messages", payload)
         self.assert_json_success(result)
 
+    def test_message_idempotency(self) -> None:
```
Let's put the test in a new `test_idempotent_request.py`.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -1455,7 +1515,11 @@ def check_send_message(
         )
     except ZephyrMessageAlreadySentError as e:
         return SentMessageResult(message_id=e.message_id)
-    return do_send_messages([message], mark_as_read=[sender.id] if read_by_sender else [])[0]
```
This `message` local variable should be called `message_request`, since it's a `SendMessageRequest`, not a `Message`. Can you do a prep commit fixing that here, and anywhere else the same mistake occurs in this file?

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -858,11 +861,59 @@ def get_active_presence_idle_user_ids(
 
 
-@transaction.atomic(savepoint=False)
+def insert_idempotent_msg(message: Message, idempotency_key: str) -> SentMessageResult | None:
```
`insert_message_with_idempotency_key` would be a better name. `msg` is an abbreviation we avoid across the codebase, and this is a lot more clear.

timabbott commented on zerver/actions/message_send.py:
```diff
@@ -879,7 +930,15 @@ def do_send_messages(
 
-    Message.objects.bulk_create(send_request.message for send_request in send_message_requests)
+    if idempotency_key is not None:
+        maybe_sent_message_result = insert_idempotent_msg(
```
Here you need to `assert len(send_message_requests) == 1`, since the code after this is wrong if that's not the case.

abdelrahman725 replied:
I would prefer to raise `JsonableError` to be more explicit saying that idempotency-key should only be expected via API request which sends only a single message

timabbott commented on zerver/tests/test_message_send.py:
```diff
@@ -1508,6 +1508,66 @@ def test_guest_user(self) -> None:
+            # indicating the response was cached successfully.
+            self.assertEqual(content["id"], first_msg_sent_id)
+            assert "automatic_new_visibility_policy" in content
+            self.assertEqual(content["automatic_new_visibility_policy"], 3)
```
I think slightly better here would be to have an deep equality check that the entire response object is identical with `isEqual` or whatever -- not just checking two fields of the object. Right now, if some extra fields like the `status` field leaked into the object when idempotent, your test would not catch the bug.

abdelrahman725 commented on web/src/util.ts:
```diff
@@ -622,3 +622,12 @@ export async function sha256_hash(text: string): Promise<string | undefined> {
 }
+
+export function generate_idempotency_key(): string | undefined {
+    // The Web Crypto API is only available in secure contexts (HTTPS or localhost).
+    if (!window.isSecureContext) {
+        return undefined;
+    }
+
+    return crypto.randomUUID();
```
Hmm, I agree. Lets fall-back to a lower-quality random UUID if we are not in a secure context.

abdelrahman725 commented on zerver/actions/message_send.py:
```diff
@@ -876,10 +926,25 @@ def do_send_messages(
+        Message.objects.bulk_create(send_request.message for send_request in send_message_requests)
+
+    else:
+        if len(send_message_requests) != 1:  # nocoverage
```
This condition can't be true for API calls anyway, so I think it's not very important to have a test case

abdelrahman725 commented on web/src/util.ts:
```diff
@@ -622,3 +622,21 @@ export async function sha256_hash(text: string): Promise<string | undefined> {
+    // The Web Crypto API is only available in secure contexts (HTTPS or localhost).
+    if (!window.isSecureContext) {
+        // A custom lower-entropy UUID V4
+        return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replaceAll(/[xy]/g, (c) => {
+            /* eslint-disable no-bitwise, unicorn/prefer-math-trunc */
+            const r = (Math.random() * 16) | 0;
+            // (r & 0x3 | 0x8) limits r to be on of (8, 9, 10, 11) only
+            // for the position y, this conforms to the UUID V4 spec.
+            const decimal_character = c === "x" ? r : (r & 0x3) | 0x8;
+            // To Hexadecimal
+            return decimal_character.toString(16);
+        });
```
This is LLM generated, BUT I understand how/why it works, tested it validating its return values against existing online validators. And all comments write written by me.

abdelrahman725 commented on zerver/actions/message_send.py:
```diff
@@ -858,11 +861,58 @@ def get_active_presence_idle_user_ids(
+    """
+
+    # Get Message db column names.
+    # These fields are filtered in the same way django bulk_create does it.
```
Relevant lines in djago implementation are [this](https://github.com/django/django/blob/main/django/db/models/query.py#L802) and [this ](https://github.com/django/django/blob/main/django/db/models/query.py#L824)

alexmv replied:
Tip for the future: Press `y` on the keyboard to make the link into a permalink -- otherwise, the line number is embedded into the link, but the code in `main` will continue to move the lines around, making the link not go where you intend when someone looks at it later/

abdelrahman725 commented:
Addressed review comments

## Changes to raw sql query logic :

Before starting the work, we `INSERT` into idempotency_table, my change is that within that same query I check if a `cached_response` already exists, and return it. We only call the decorated function if there is no `cached_response`. 

`do_idempotent_insert` is now only responsible for inserting (without checking duplicates) -- yes it's safe, see my testing below--- and it keeps `FOR UPDATE NOWAIT` for locking.

### Why that change ?
Because I believe it's much readable and simpler, also this would avoid us having to change the signature of the decorated function which would have to return another variable (e.g. duplicate) to check it and set the `cached_response` accordingly.

### Testing for correctness

I tested aggressively this time:

1. Sent `15` concurrent requests with the same `idempotency-key`, only one message is created, getting some few `409`, rest of responses are identical as expected.

2. Sent `24` concurrent requests with `6` **different**  `idempotency-key`s, in a round-robin style, results are as expected and only 6 messages are created.

showell commented on zerver/lib/idempotent_request.py:
```diff
@@ -0,0 +1,180 @@
+                            VALUES (%(realm_id)s, %(user_id)s, %(idempotency_key_value)s)
+                            ON CONFLICT DO NOTHING;
+
+                            SELECT cached_response AS cached_response FROM {idempotency_table}
```
I don't understand the need to say "cached_response AS cached_response", but maybe that's just my ignorance of SQL.

alexmv replied:
I know of no reason to have an explicit `AS` here.

showell commented on zerver/middleware.py:
```diff
@@ -750,3 +758,18 @@ def process_request(self, request: HttpRequest) -> HttpResponse | None:
+
+
+# We can also convert this middleware
+# into a generic one to validate other headers as well.
```
I think a more useful comment here would be to just explain what this particular function does.

I also think it's useful to explain that this particular function only exists in the middleware to prevent obviously bad UUIDs at the middleware layer (if I am reading this correctly).  Presumably most of the logic happens outside of the middleware, since the "Idempotency-Key" header presumably has to have some value that matches up with what's in the database or whatever.

showell commented on zerver/models/idempotent_requests.py:
```diff
@@ -0,0 +1,44 @@
+
+    # Rows with timestamp later than this should be deleted.
+    # TODO: Should be used by a cron job.
+    EXPIRE_AFTER = 60 * 60  # Time in seconds.
```
If this isn't yet enforced by anything, then I would just make the whole thing by a TODO comment that says something like "# TODO: make a cron job that expires keys after an hour."

alexmv replied:
Agree.  My previous comment was intended to imply "we should remove this constant, or write the code that uses it."  Having unused constants isn't helpful -- adding a TODO comment just makes them slightly more verbosely unhelpful. :)

alexmv commented on zerver/lib/idempotent_request.py:
```diff
@@ -0,0 +1,180 @@
+    destination_model: type[models.Model],
+    field_names: list[str],
+    column_to_value: dict[str, Any],
+) -> DictRow:
```
This is still returning information which is tied to your underlying implementation.  Why doesn't this return `int`?  Why make callers dig in an extract `new_id`?  This also disagrees with the stated return type on line 34, below.

abdelrahman725 replied:
Yes, I agree

alexmv commented on zerver/models/idempotent_requests.py:
```diff
@@ -0,0 +1,44 @@
+    # Not Null: work is done and succeeded/failed.
+    identifier = models.IntegerField(null=True, db_index=True)
+
+    # Currently, this is the cached result of the work, not the json response itself.
+    cached_response = models.JSONField(default=None, null=True)
```
If the name is currently a lie, we should probably adjust the name.

alexmv commented on zerver/lib/idempotent_request.py:
```diff
@@ -0,0 +1,180 @@
+    durable: bool = False,
+    savepoint: bool = True,
+    *,
+    cached_response_serializer: Callable[[Any], dict[str, Any]],
+    cached_response_deserializer: Callable[[str], Any],
```
The fact that these are not inverses of each other is a smell.  Specifically, the `cached_response` database column is a JSONB column, so we _know_ it's not just going to be random string.  Every `cached_response_deserializer` is going to need to call `orjson.loads`.  We can, and should, do that JSON decoding in `idempotent_request_transaction` for them, and pass the deserializer a `dict[str, Any]` -- which will make the types of these two functions nicely inverses of each other, as they should be.

alexmv commented on zerver/lib/idempotent_request.py:
```diff
@@ -0,0 +1,180 @@
+    column_to_value: dict[str, Any],
+) -> DictRow:
+    """
+    Important: MUST run inside idempotent_request_transaction.
```
Did you research if there was a way to enforce this programmatically?

abdelrahman725 commented on zerver/lib/idempotent_request.py:
```diff
@@ -0,0 +1,212 @@
+                    ).update(completed=True, cached_result=cached_result_serializer(result))
+                    return result
+
+                # Row is locked by another concurrent request doing the work.
+                except OperationalError as error:  # nocoverage
+                    # We don't test this, because it's raised only during TRUE concurrent requests
+                    # which can't be reproduced in zulip tests.
```
**Important**

I couldn't reproduce true concurrency within Zulip tests, I tried mock approach and asyncio, but neither of these raised `LockNotAvailable` error.

The only way I could achieve true concurrency is using the following external (outside zulip dev environment) python script, with very small and negligible time difference of less than `1ms` between each adjacent request.

That's why I add `# nocoverage`, as I beleive those tests wouldn't be beneficial.

I also used another version of the script to send `40` concurrent requests **BUT using ONLY one** `Idempotency-Key`  value.

Script used:
```python3
import time
import asyncio
import aiohttp

headers = {
    "Host": "localhost:9991",
    "sec-ch-ua": '"Not?A_Brand";v="8", "Chromium";v="108"',
    "sec-ch-ua-mobile": "?0",
    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/108.0.5359.125 Safari/537.36",
    "Content-Type": "application/x-www-form-urlencoded; charset=UTF-8",
    "Accept": "application/json, text/javascript, */*; q=0.01",
    "X-Requested-With": "XMLHttpRequest",
    "X-CSRFToken": "qARD6pAgfb7OkpnhDWXe1BHlMNtsdm3xjdFrC5tkYwVO3JjN8J1W6dJsu8tROFJy",
    "sec-ch-ua-platform": '"Linux"',
    "Origin": "http://localhost:9991",
    "Sec-Fetch-Site": "same-origin",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Dest": "empty",
    "Referer": "http://localhost:9991/",
    "Accept-Encoding": "gzip, deflate",
    "Accept-Language": "en-US,en;q=0.9",
    "Cookie": "csrftoken=3NYYGQ3eTvYaTu6GFXeSfMchSvazLtQb; sessionid=rz0s8ssdigdq0kf3z76xjlgdkaqybs3x; django_language=en",
    "Connection": "close",
}

url = "http://localhost:9991/json/messages"

data = (
    "type=stream&content=MyMessage&sender_id=9&queue_id=c74f244e-41de-4cc2-ba24-8f6fad8dd86e&"
    "topic=Idempotency&stream_id=4&to=%5B4%5D&draft_id=199aca24e12-b82970c203de8&"
    "local_id=194.01&locally_echoed=true&resend=false"
)

idempotency_keys = [
    "8e118d41-39d7-4a82-8da9-a6a3162d5700",
    "8e118d41-39d7-4a82-8da9-a6a3162d5700",
    "8e118d41-39d7-4a82-8da9-a6a3162d5700",
    "8e118d41-39d7-4a82-8da9-a6a3162d5700",

    "8e118d41-39d7-4a82-8da9-a6a3162d5701",
    "8e118d41-39d7-4a82-8da9-a6a3162d5701",
    "8e118d41-39d7-4a82-8da9-a6a3162d5701",
    "8e118d41-39d7-4a82-8da9-a6a3162d5701",

    "8e118d41-39d7-4a82-8da9-a6a3162d5702",
    "8e118d41-39d7-4a82-8da9-a6a3162d5702",
    "8e118d41-39d7-4a82-8da9-a6a3162d5702",
    "8e118d41-39d7-4a82-8da9-a6a3162d5702",
    
    "8e118d41-39d7-4a82-8da9-a6a3162d5703",
    "8e118d41-39d7-4a82-8da9-a6a3162d5703",
    "8e118d41-39d7-4a82-8da9-a6a3162d5703",

    "8e118d41-39d7-4a82-8da9-a6a3162d5704",
    "8e118d41-39d7-4a82-8da9-a6a3162d5704",
    "8e118d41-39d7-4a82-8da9-a6a3162d5704",

    "8e118d41-39d7-4a82-8da9-a6a3162d5705",
    "8e118d41-39d7-4a82-8da9-a6a3162d5705",
    "8e118d41-39d7-4a82-8da9-a6a3162d5705",

    "8e118d41-39d7-4a82-8da9-a6a3162d5706",
    "8e118d41-39d7-4a82-8da9-a6a3162d5706",
    "8e118d41-39d7-4a82-8da9-a6a3162d5706",

    "8e118d41-39d7-4a82-8da9-a6a3162d5707",
    "8e118d41-39d7-4a82-8da9-a6a3162d5707",
    "8e118d41-39d7-4a82-8da9-a6a3162d5707",

    "8e118d41-39d7-4a82-8da9-a6a3162d5708",
    "8e118d41-39d7-4a82-8da9-a6a3162d5708",
    "8e118d41-39d7-4a82-8da9-a6a3162d5708",

    "8e118d41-39d7-4a82-8da9-a6a3162d5709",
    "8e118d41-39d7-4a82-8da9-a6a3162d5709",
    "8e118d41-39d7-4a82-8da9-a6a3162d5709",       
]

async def async_send_message(session, i, key):
    headers["Idempotency-Key"] =  key
    print(f"Sending message at: {time.time():.6f}")
    async with session.post(url, headers=headers, data=data, timeout=5) as response:
        body = await response.text()
        print(f"Req No. {i+1} Status: {response.status} | Body: {body}")


async def main():
    async with aiohttp.ClientSession() as session:
        tasks = [async_send_message(session, i, key) for i, key in enumerate(idempotency_keys)]
        await asyncio.gather(*tasks)
```
